Mathlib/Data/UInt.lean

```diff
@@ -1,10 +1,231 @@
 
-theorem Fin.val_eq_of_lt : a < n.succ → (@Fin.ofNat n a).val = a := Nat.mod_eq_of_lt
+lemma UInt8.size_positive : 0 < UInt8.size := by decide
 
-theorem UInt32.val_eq_of_lt : a < UInt32.size → (UInt32.ofNat a).val = a := Fin.val_eq_of_lt
+lemma UInt16.size_positive : 0 < UInt16.size := by decide
+
+lemma UInt32.size_positive : 0 < UInt32.size := by decide
+
+lemma UInt64.size_positive : 0 < UInt64.size := by decide
+
+lemma USize.size_positive : 0 < USize.size := by
+  simp only [USize.size]
+  cases System.Platform.numBits_eq with
+  | inl eq32 => simp [eq32]
+  | inr eq64 => simp [eq64]
+
+set_option hygiene false
+macro "genIntDeclars" typeName:ident : command => do
+  `(
+    namespace $typeName
+      instance : Zero (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : One (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : Neg (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : AddSemigroup (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : AddMonoid (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : AddCommMonoid (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : SubNegMonoid (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : AddGroup (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : Monoid (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : CommMonoid (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      instance : MonoidWithZero (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
+
+      lemma val_eq_of_lt : a < size → (ofNat a).val = a := Fin.val_eq_of_lt
+
+      lemma sub_def (a b : $typeName) : a - b = ⟨a.val - b.val⟩ := rfl
+
+      lemma mul_def (a b : $typeName) : a * b = ⟨a.val * b.val⟩ := rfl
+
+      lemma mod_def (a b : $typeName) : a % b = ⟨a.val % b.val⟩ := rfl
+
+      lemma add_def (a b : $typeName) : a + b = ⟨a.val + b.val⟩ := rfl
+
+      lemma eq_of_val_eq : ∀ {a b : $typeName}, a.val = b.val -> a = b
+      | ⟨f1⟩, ⟨f2⟩, h => congrArg mk h
+
+      lemma val_eq_of_eq : ∀ {a b : $typeName}, a = b -> a.val = b.val
+      | ⟨f1⟩, ⟨f2⟩, h => congrArg val h
+
+      @[simp] lemma mk_val_eq (a : $typeName) : mk a.val = a := by
+        cases h:a with
+        | mk aFin =>
+          have _ : a.val = aFin := h ▸ rfl
+          simp only [‹a.val = aFin›]
+
+      lemma add_assoc (a b c : $typeName) : (a + b) + c = a + (b + c) :=
+        congrArg mk (Fin.add_assoc _ _ _)
+
+      lemma mul_comm (a b : $typeName) : a * b = b * a := congrArg mk (Fin.mul_comm _ _)
+
+      lemma add_comm (a b : $typeName) : a + b = b + a := congrArg mk (Fin.add_comm _ _)
+
+      @[simp] lemma add_zero (a : $typeName) : a + 0 = a := by
```

Comment:
Can you inline these lemmas into the instances?  There is no reason to have separate lemmas for UInt* when we already have the same lemma for AddMonoid.

If you can't inline it, please make it at least `protected` (so that it doesn't pollute the namespace with `open UInt*`) and remove the `@[simp]`.